### PR TITLE
Couchforms analytics

### PR DESCRIPTION
**Pull request by dannyroberts**
@czue waiting for tests

There's lots more of these to go, but probably no harm merging this in first

### CODE REVIEW

**Comment by czue** [comment]
looks like a great set of changes. I think I'll feel much more comfortable if we have tests that exercise these functions. once those exist :shipit:

File: corehq/ex-submodules/couchforms/analytics.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    from corehq.apps.reports.util import make_form_couch_key
+    key = make_form_couch_key(None)
+    return {
+        row["key"][1]: row["value"]
+        for row in XFormInstance.get_db().view(
+            "reports_forms/all_forms",
+            group=True,
+            group_level=2,
+            startkey=key,
+            endkey=key+[{}],
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/analytics.py]: E226 missing whitespace around arithmetic operator

File: corehq/ex-submodules/couchforms/analytics.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+
+def get_number_of_forms_in_domain(domain):
+    from corehq.apps.reports.util import make_form_couch_key
+    key = make_form_couch_key(domain)
+    row = XFormInstance.get_db().view(
+        "reports_forms/all_forms",
+        startkey=key,
+        endkey=key+[{}],
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/analytics.py]: E226 missing whitespace around arithmetic operator

File: corehq/ex-submodules/couchforms/tests/test_analytics.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    get_first_form_submission_received, get_last_form_submission_received, \
+    app_has_been_submitted_to_in_last_30_days, update_analytics_indexes
+from couchforms.models import XFormInstance
+
+
+class CouchformsAnalyticsTest(TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        from casexml.apps.case.tests import delete_all_xforms
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/tests/test_analytics.py]: F401 'delete_all_xforms' imported but unused

File: corehq/apps/ota/views.py [hunk 4]
```diff
@@ -14,6 +14,7 @@
 from corehq.apps.style.views import BaseB3SectionPageView
 from corehq.apps.users.models import CouchUser, CommCareUser
 from corehq.util.view_utils import json_error
+from couchforms.dbaccessors import get_form_ids_for_user, get_form_xml_element
```
Comment (dimagimon) [inline corehq/apps/ota/views.py]: F401 'get_form_ids_for_user' imported but unused

Comment (dimagimon) [inline corehq/apps/ota/views.py]: F401 'get_form_xml_element' imported but unused

**Comment by dimagimon** [comment]
There are 1 errors:

* corehq/apps/commtrack/tests/util.py, line 12 - F401 'get_commtrack_forms' imported but unused

**Comment by dimagimon** [comment]
There are 4 errors:

* corehq/apps/commtrack/tests/util.py, line 12 - F401 'get_commtrack_forms' imported but unused
* corehq/ex-submodules/couchforms/analytics.py, line 2 - F401 'get_db' imported but unused
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 26 - E501 line too long (125 > 115 characters)
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 28 - E501 line too long (123 > 115 characters)

**Comment by dimagimon** [comment]
There are 4 errors:

* corehq/apps/commtrack/tests/util.py, line 12 - F401 'get_commtrack_forms' imported but unused
* corehq/ex-submodules/couchforms/analytics.py, line 2 - F401 'get_db' imported but unused
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 26 - E501 line too long (125 > 115 characters)
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 28 - E501 line too long (123 > 115 characters)

**Comment by dimagimon** [comment]
There are 4 errors:

* corehq/apps/commtrack/tests/util.py, line 12 - F401 'get_commtrack_forms' imported but unused
* corehq/ex-submodules/couchforms/analytics.py, line 2 - F401 'get_db' imported but unused
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 26 - E501 line too long (125 > 115 characters)
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 28 - E501 line too long (123 > 115 characters)

File: corehq/ex-submodules/couchforms/analytics.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+
+def get_last_form_submission_for_user_for_app(domain, user_id, app_id=None):
+    if app_id:
+        key = ['submission app user', domain, app_id, user_id]
+    else:
+        key = ['submission user', domain, user_id]
+    xform = XFormInstance.view(
+        "reports_forms/all_forms",
+        startkey=key+[{}],
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/analytics.py]: E226 missing whitespace around arithmetic operator

**Comment by dannyroberts** [comment]
:white_check_mark:

**Comment by dannyroberts** [comment]
@czue thanks for suggesting tests. Adding them was super easy and super helpful.

**Comment by czue** [comment]
looks like a great set of changes. reviewed individual commits for structure though didn't review the functions closely. I think I'm comfortable with that level given that it was mostly moved code and that it now has test coverage too (!!). will merge assuming you agree with that sentiment.

one question - wanted to confirm that you were careful about the convention of `analytics` being something that is only used in reports? so like, we can now rewrite those functions on ES or sofabed with no major concerns. right?

assuming yes is probably worth highlighting the convention to the wider team

**Comment by dannyroberts** [comment]
@czue yeah, that's right.

@dimagi/team-commcare-hq I'm setting a new convention in this PR, most relevant to our bulkier data models i.e. forms & cases, of splitting out `analytics` accessors from `dbaccessors`. The idea is that `dbaccessors` are what's needed for core operations and must be always up-to-date and exactly correct, whereas `analytics` can and should use `stale=stale_ok()` if querying couch or can be rewritten to work off ES or sofabed, because they're used for reports and analytics and other read-heavy applications where exact up-to-dateness is not a hard requirement.

**Comment by dannyroberts** [comment]
As a bonus, this PR removes the last usages of the following couch views:

- `reports_forms/all_submitted_users`
- `hqadmin/domains_over_time`

which we can remove at any time here: https://github.com/dimagi/commcare-hq/pull/8341 (but will require a pretty large reindex)

**Pull request merged**